Wrap mEstimators

varunagrawal opened the pull request:
1. Wrap more mEstimators.
2. Also wrap the `weight` function of each of the mEstimators.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/borglab/gtsam/117)
<!-- Reviewable:end -->

dellaert commented:
Can we add the gtsam.h changes and the matlab script in here, as well?

varunagrawal commented:
This PR builds upon #116. Once 116 is approved, this one will only show changes to `gtsam.h`.

varunagrawal commented:
This PR no longer depends on #116.

dellaert commented on gtsam.h:
```diff
@@ -1378,6 +1382,18 @@ virtual class Huber: gtsam::noiseModel::mEstimator::Base {
+  double weight(double error) const;
+};
+
+virtual class Cauchy: gtsam::noiseModel::mEstimator::Base {
```
You know, inheritance works in gtsam.h. Can you not just add weight to the Base?

dellaert replied:
Limitation of wrap, understood.

dellaert commented on matlab/gtsam_examples/VisualizeMEstimators.m:
```diff
@@ -0,0 +1,172 @@
+import gtsam.*
```
Please add documentation and comments? Take a look at LocalizationExample.m to see the documentation standards. In addition, could we scale all figures the same way so the models are directly comparable?

dellaert commented on matlab/gtsam_examples/VisualizeMEstimators.m:
```diff
@@ -0,0 +1,172 @@
+    rho = c^2 * ( (abs(x) ./ c) - log(1 + (abs(x)./c)) );
+    est = noiseModel.mEstimator.Fair(c);
+    
+    w = zeros(size(x));
```
Copy/paste

ProfFan replied:
Is this resolved?

varunagrawal replied:
Yup

dellaert commented on matlab/gtsam_examples/VisualizeMEstimators.m:
```diff
@@ -0,0 +1,172 @@
+    import gtsam.*
+    c = 1.3998;
+
+    rho = c^2 * ( (abs(x) ./ c) - log(1 + (abs(x)./c)) );
```
So, this is crazy. We should, as Mike Bosse suggested, put the residuals into the noiseModel header. Let's talk if we can then kill two birds with one stone and immediately understand and fix the problem he is talking about.

varunagrawal replied:
This is done. Residual functions are implemented and wrapped.

dellaert commented on matlab/gtsam_examples/VisualizeMEstimators.m:
```diff
@@ -0,0 +1,172 @@
+plot_w(x, w, 3, 2);
+saveas(figure(6), 'tukey.png');
+
+function plot_rho(x, y, idx, yll, ylu)
```
Can we just combine those three functions? Docs...

ProfFan commented:
#121

varunagrawal commented:
![tukey](https://user-images.githubusercontent.com/975964/65646855-82a8d000-dfca-11e9-8b12-12244a6cbd20.png)
![welsch](https://user-images.githubusercontent.com/975964/65646856-82a8d000-dfca-11e9-891b-b31470fe1c41.png)
![gemanmcclure](https://user-images.githubusercontent.com/975964/65646857-82a8d000-dfca-11e9-83a7-42d2535dfac1.png)
![cauchy](https://user-images.githubusercontent.com/975964/65646858-82a8d000-dfca-11e9-824c-28538a4761c4.png)
![huber](https://user-images.githubusercontent.com/975964/65646859-82a8d000-dfca-11e9-8f4d-6909fc94a131.png)
![fair](https://user-images.githubusercontent.com/975964/65646860-82a8d000-dfca-11e9-890a-1c5081fd9765.png)

varunagrawal commented:
I need to regenerate Huber with the fixed code. :(

dellaert commented:
I guess rebase this on develop should fix it?

dellaert commented:
Do we now have the residuals into the noiseModel header?

varunagrawal commented:
We don't. Should I prioritize that over other tasks?

dellaert commented:
The blog post is priority. I thought this PR was part of it. I guess you can do without, but we can't land this PR without addressing this. You could fix this after the blog post.

varunagrawal commented:
I have added the residuals. However, I can't find the code that calls the weight function to compute the total penalty. That would be the last thing to update (along with the Matlab example) before landing this PR.

dellaert commented on gtsam.h:
```diff
@@ -1370,6 +1373,8 @@ virtual class Fair: gtsam::noiseModel::mEstimator::Base {
   // enabling serialization functionality
   void serializable() const;
+
+  double weight(double error) const;
```
No residual?

varunagrawal replied:
Yup working on that.

dellaert commented on gtsam.h:
```diff
@@ -1394,8 +1416,23 @@ virtual class Welsch: gtsam::noiseModel::mEstimator::Base {
+  double residual(double error) const;
 };
 
+//TODO DCS and L2WithDeadZone mEstimators
```
Do it?

varunagrawal replied:
Done!

dellaert commented on gtsam/linear/NoiseModel.h:
```diff
@@ -18,6 +18,7 @@
 
 #pragma once
 
+#include<iostream>
```
Nonono! Iosfwd is *reluctantly* tolerated, iostream is not. We want as few headers included in .h files as possible, to cut down on compile time. Especially a header like noiseModel.

Also, header order in GTSAM is near to far, so iosfwd (IF needed) would come last.

varunagrawal replied:
Woops. I was using that for debugging and accidentally left it in. :disappointed:

dellaert commented on gtsam/linear/NoiseModel.h:
```diff
@@ -752,6 +754,12 @@ namespace gtsam {
           return 1.0 / (1.0 + std::abs(error) / c_);
         }
+        double residual(double error) const {
+          double absError = std::abs(error);
```
Probably compiler will figure it out, but add const to all these?

varunagrawal replied:
Yup added to these as well as function definitions for other M-estimators.

dellaert commented on gtsam/linear/NoiseModel.h:
```diff
@@ -779,6 +787,14 @@ namespace gtsam {
         }
+        double residual(double error) const {
+          double absError = std::abs(error);
+          if (absError <= k_) {  // |x| <= k
```
Not: above it is < not <=

varunagrawal replied:
Actually, as per the definition of Huber loss, it should be `<=` for both.
https://en.wikipedia.org/wiki/Huber_loss#Definition

dellaert commented on gtsam/linear/NoiseModel.h:
```diff
@@ -839,6 +860,16 @@ namespace gtsam {
+          if (absError <= c_) {
+            double t = csquared_ - (error*error);
+            double val = t * t * t;
+            return ((csquared_*csquared_*csquared_) - val) / (6.0 * csquared_ * csquared_);
```
So, c^6 ?? Seems weird. Is this validated by a test? Also, should be computed in constructor then.

varunagrawal replied:
I modified this to make this easier to read and correlate with the actual expression.

dellaert commented on gtsam/linear/NoiseModel.h:
```diff
@@ -897,6 +932,10 @@ namespace gtsam {
           double xc2 = (error*error)/csquared_;
           return std::exp(-xc2);
         }
+        double residual(double error) const {
```
Is pretty horrible copy/paste. can we not just say `using Welsh = Welsch’ ??

varunagrawal replied:
Yeah I would actually prefer that. This is jarring.

dellaert commented on gtsam/linear/NoiseModel.h:
```diff
@@ -751,15 +750,8 @@ namespace gtsam {
-          double val = normalizedError - std::log(1 + normalizedError);
-          return c_ * c_ * val;
-        }
+        double weight(double error) const;
```
Const then? But, I normally consider const redundant on arguments passed by value...

varunagrawal replied:
Yeah I was wondering what your policy on that was? Some people say const in the declaration is redundant since it doesn't do anything. Should I remove the const for the function arguments then? I felt it was a nice built-in check since the values should not change.

dellaert replied:
I would say undo these changes. We don’t do it anywhere else in GTSAM, and it is orthogonal to the goal of the PR

varunagrawal replied:
Understood.

varunagrawal commented on gtsam/linear/NoiseModel.h:
```diff
@@ -924,7 +890,8 @@ namespace gtsam {
         GemanMcClure(double c = 1.0, const ReweightScheme reweight = Block);
         virtual ~GemanMcClure() {}
-        virtual double weight(double error) const;
+        virtual double weight(const double error) const;
```
@dellaert what do you have to say about these `virtual` identifiers? Would we need to subclass the estimators?

dellaert replied:
Good point: the base class should have virtual, but derived classes should have override in c++11

varunagrawal replied:
Doesn't this mean all the estimators should use `override`?

dellaert replied:
Yes

varunagrawal replied:
All done.

ProfFan commented on gtsam/linear/NoiseModel.cpp:
```diff
@@ -718,15 +718,25 @@ void Null::print(const std::string &s="") const
-/* ************************************************************************* */
+double Fair::weight(double error) const {
+  return 1.0 / (1.0 + std::abs(error) / c_);
+}
```
Newline?

ProfFan commented on gtsam/linear/NoiseModel.cpp:
```diff
@@ -791,7 +825,30 @@ Cauchy::shared_ptr Cauchy::Create(double c, const ReweightScheme reweight) {
+    return (1.0-xc2)*(1.0-xc2);
+  }
+  return 0.0;
+}
```
Same

ProfFan commented on gtsam/linear/NoiseModel.h:
```diff
@@ -727,6 +727,7 @@ namespace gtsam {
         Null(const ReweightScheme reweight = Block) : Base(reweight) {}
         virtual ~Null() {}
```
Why some of the destructors are virtual and some not?

varunagrawal replied:
Oops I missed that. Fixed.

varunagrawal commented:
@dellaert @ProfFan the CI should now pass and I have fixed a few more things (such as adding `residual` wrapper to Fair).

I have also gone ahead and addressed Mike Bosse's TODOs, and hence I am asking for a review once again to ensure I am doing the right thing.

dellaert commented on gtsam/linear/NoiseModel.h:
```diff
@@ -1069,8 +1018,7 @@ namespace gtsam {
       inline virtual double distance(const Vector& v) const
-      { return this->whiten(v).squaredNorm(); }
-      // TODO(mike): fold the use of the m-estimator residual(...) function into distance(...)
+      { return robust_->residual(v.norm()); }
```
This is suspect. Where did the whiten go. If unit tests did not catch this, it means a unit test is missing :-) We should probably add unit tests for regular and some robust model, e.g. Huber.

ProfFan replied:
2nd

varunagrawal replied:
So there was a unit test which was commented out. It had the same TODO note as this issue, so I uncommented it and made the changes, and it passed without issues. :slightly_smiling_face: 

But yes, I was on the fence wrt whether this should be a new PR. Thanks for clarifying this.

varunagrawal commented:
@dellaert undid the last commit. Will make a new PR with the changes. Mike Bosse is not a collaborator on Github, so we'll have to ask him separately to help out on this.

dellaert commented:
Merge at will

michaelbosse commented on gtsam/linear/NoiseModel.cpp:
```diff
@@ -774,6 +799,16 @@ Cauchy::Cauchy(double k, const ReweightScheme reweight) : Base(reweight), k_(k),
+}
+
+double Cauchy::residual(double error) const {
+  const double xc2 = error / k_;
```
minor nit: `xc2` is usually `error*error/ksquared_`, and then can do `std::log(1 + xc2)` below.

dellaert replied:
@varunagrawal I know you can't finish this yet because of other priorities, but could you incorporate this change into the new PR?
